### ISSUE TITLE
TextField HelpText element loses id attribute when switching from description to errorMessage

### ISSUE THREAD

**Issue by majornista** [adobe/react-spectrum#2330]
# 🐛 Bug Report 

In the [TextField HelpText example](https://react-spectrum.adobe.com/react-spectrum/TextField.html#help-text), and any examples where the HelpText toggles between `descriptionProps` and `errorMessageProps`. The HelpText initializes with the `id` attribute present, but once the validation state changes and the description is swapped for the `errorMessage`, the `id` on the HelpText goes away, and the `aria-describedby` on the input is no longer valid.

## 🤔 Expected Behavior

`aria-describedby` on the `input` should include a reference to the description and/or the errorMessage slots by `id`.

## 😯 Current Behavior

If we display only the description or the errorMessage using HelpText, only one slot will be rendered to the DOM, and the second `id` created with `useSlotId` gets left on the cutting room floor. When we only render one slot, the HelpText should use whichever `id` renders first.

## 💁 Possible Solution

At https://github.com/adobe/react-spectrum/blob/058ce77164ac2b5fa9332634253d395d9c382ac4/packages/%40react-aria/label/src/useField.ts#L50-L55,
we can use: 
```tsx
  if (description) {
    descriptionProps.id = descriptionId || errorMessageId;
  }
  if (errorMessage) {
    errorMessageProps.id = errorMessageId || descriptionId;
  }
```
which ensures that if one of the slots is not defined, the HelpText element will receive an `id`.

## 🔦 Context

Discovered while testing [TextField HelpText example](https://react-spectrum.adobe.com/react-spectrum/TextField.html#help-text) using VoiceOver on macOS.

## 💻 Code Sample
```tsx
function Example() {
  let [value, setValue] = React.useState('0');
  let isValid = React.useMemo(() => /^\d$/.test(value), [value]);

  return (
    <TextField
      validationState={isValid ? 'valid' : 'invalid'}
      value={value}
      onChange={setValue}
      label="Favorite number"
      maxLength={1}
      description="Enter a single digit number."
      errorMessage={
        value === ''
          ? 'Empty input not allowed.'
          : 'Single digit numbers are 0-9.'
      }
    />
  );
}
```
## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   |  3.2.0
| Browser          |  Chrome 93.0.4577.63 (Official Build) (x86_64)
| Operating System | MacOS Big Sur version 11.5.2 (20G95)

## 🧢 Your Company/Team

Adobe/Accessibility

**Comment by snowystinger** [comment]
Been asked to record thoughts here as well. It's similar to what's in the PR already, but this is an easier discussion forum I think.

The issue is in useSlotId. It only validates the id on mount. Subsequent rerender that may remove of add children do not cause re-renders. There are a few different approaches we could use to handle this.

As mentioned in the PR, useId/useSlotId doesn't store our id map in state, so moving to state is probably something we should be doing anyways. https://github.com/Light-Keeper/react-singleton-hook
I don't think this will solve our issue though, because that still wouldn't alert us when children are added/removed. So I think we can table this discussion for another time.

So I see a few ways we could approach the issue of children removed/added.
- We could have useSlotId send back a mount/unmount function (maybe a hook?). This would be breaking, so maybe don't do this.
- We could add a mutation observer looking for any children of the useSlotId with the id either added or removed. When they appear/disappear, we update https://github.com/adobe/react-spectrum/blob/main/packages/@react-aria/utils/src/useId.ts#L93 and thereby trigger a rerender with the updated Id. We'd need to add an optional ref as an argument to useSlotId so it's not breaking. If you don't provide a ref, then it behaves as it has and only checks on mount, otherwise observes. This seems like a good option.
- We check on every render. We may have children with their own rendering that doesn't cause the component with useSlotId to re-render, so this is probably not a good route to go. Though it may be the fastest route to fix the current issue.

**Comment by snowystinger** [comment]
Adding more notes now that we looked into it a bit more
While the commit https://github.com/adobe/react-spectrum/pull/2332/commits/6cf53a187bc769b98e238ebeacf18441ce1efc73 does cause HelpText to function correctly, I do have a concern with the approach:
- It creates a new hook, we don't want that, and there isn't a way to do it without modifying the current hook's return types
- If the components are user provided, they may override the ref we pass along on context

We tried the mutation observer approach as well, however, it requires a ref to some sort of container, and it requires that that ref be passed not just through useSlotId, but also useField and useTextfield and probably some of the others that use useField. In addition, the container may be outside the components control, such as in one of the considerations for the rewrite of forms where the form itself would be the parent to everything, labels, inputs and descriptions/errors.
